server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/UndesiredAllocationsTracker.java

```diff
@@ -0,0 +1,179 @@
+import org.elasticsearch.logging.LogManager;
+import org.elasticsearch.logging.Logger;
+
+/**
+ * Keeps track of a limited number of shards that are currently in undesired allocations. If the
+ * shards remain in undesired allocations for longer than a configurable threshold, it will log
+ * why those shards can't be allocated to desired nodes.
+ */
+public class UndesiredAllocationsTracker {
+
+    private static final Logger logger = LogManager.getLogger(UndesiredAllocationsTracker.class);
+
+    /**
+     * Warning logs will be periodically written if we see a shard that's been in an undesired allocation for this long
+     */
+    public static final Setting<TimeValue> UNDESIRED_ALLOCATION_DURATION_LOG_THRESHOLD_SETTING = Setting.timeSetting(
+        "cluster.routing.allocation.desired_balance.undesired_duration_logging.threshold",
+        TimeValue.timeValueMinutes(5),
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+
+    /**
+     * The max number of undesired allocations to track. We expect this to be relatively small.
+     */
+    public static final Setting<Integer> MAX_UNDESIRED_ALLOCATIONS_TO_TRACK = Setting.intSetting(
+        "cluster.routing.allocation.desired_balance.undesired_duration_logging.max_to_track",
+        10,
+        0,
+        200,
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+
+    private final TimeProvider timeProvider;
+    private final ObjectLongHashMap<ShardRouting> undesiredAllocations = new ObjectLongHashMap<>();
+    private final FrequencyCappedAction undesiredAllocationDurationLogInterval;
+    private volatile TimeValue undesiredAllocationDurationLoggingThreshold;
+    private volatile int maxUndesiredAllocationsToTrack;
+
+    UndesiredAllocationsTracker(ClusterSettings clusterSettings, TimeProvider timeProvider) {
+        this.timeProvider = timeProvider;
+        this.undesiredAllocationDurationLogInterval = new FrequencyCappedAction(timeProvider::relativeTimeInMillis, TimeValue.ZERO);
+        clusterSettings.initializeAndWatch(
+            DesiredBalanceReconciler.UNDESIRED_ALLOCATIONS_LOG_INTERVAL_SETTING,
+            undesiredAllocationDurationLogInterval::setMinInterval
+        );
+        clusterSettings.initializeAndWatch(
+            UNDESIRED_ALLOCATION_DURATION_LOG_THRESHOLD_SETTING,
+            value -> undesiredAllocationDurationLoggingThreshold = value
+        );
+        clusterSettings.initializeAndWatch(MAX_UNDESIRED_ALLOCATIONS_TO_TRACK, value -> this.maxUndesiredAllocationsToTrack = value);
+    }
+
+    /**
+     * Track an allocation as being undesired
+     */
+    public void trackUndesiredAllocation(ShardRouting shardRouting) {
+        assert shardRouting.unassigned() == false : "Shouldn't record unassigned shards as undesired allocations";
+        if (undesiredAllocations.containsKey(shardRouting) == false && undesiredAllocations.size() < maxUndesiredAllocationsToTrack) {
```

Comment:
Nit: maybe check size first which is faster when the tracking is full.

Reply:
Done in 0af805059ba298e4f11bc99a3805af93d9ee5375